engine/notifications.js

```diff
@@ -0,0 +1,298 @@
+  .catch(err => {
+    if (err.message !== 'Not Found') {
+      throw err;
+    }
+    return [];
+  })
+  .then(regFeatures => {
+    if (!doNotConfirm && endpoint) {
+      return sendConfirmation(endpoint, key, deviceId, regFeatures, dbNumber)
+      .then(() => regFeatures);
+    }
+    return regFeatures;
+  });
+}
+
+// Registers to all notifications including new ones
+// Removes registrations to individual features
+function registerToAll(deviceId, dbNumber) {
+  return getRegisteredFeatures(deviceId, dbNumber)
+  .then(features => unregister(deviceId, features, dbNumber, true))
+  .then(() => Promise.all([
+    redis.sadd(client, 'all-notifications', deviceId),
+    redis.sadd(client, `${deviceId}-notifications`, 'all')]));
+}
+
+// registers to receive notifications
+// required info:
+// * deviceId
+// * list of feature slugs to register to
+//   special features are `new` and `all`
+//   * new - do notify about new features
+//   * all - all features including new, registering to all will remove
+//     all individual registrations
+// * endpoint [optional]
+// one can register to a feature without providing an endpoint
+// if it was already saved:
+// * key
+// * dbNumber
+function register(deviceId, features, endpoint, key, dbNumber) {
+  if (!features || features.length === 0) {
+    return Promise.reject(new Error('No features provided'));
+  }
+
+  return setClient(dbNumber)
+  .then(() => {
+    if (endpoint) {
+      const device = {
+        id: deviceId,
+        endpoint,
+        key: key || '',
+      };
+      redis.hmset(client, `device-${deviceId}`, device);
+      return device;
+    }
+    return redis.hgetall(client, `device-${deviceId}`);
+  })
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('No endpoint provided');
+    }
+  })
+  .then(() => {
+    if (features.indexOf('all') >= 0) {
+      return registerToAll(deviceId, dbNumber);
+    }
+    return Promise.all(
+      features.map(slug => [
+        redis.sadd(client, `${slug}-notifications`, deviceId),
+        redis.sadd(client, `${deviceId}-notifications`, slug)]));
+  })
+  .then(() => sendConfirmationToDevice(deviceId, dbNumber))
+  .then(() => getRegisteredFeatures(deviceId, dbNumber));
+}
+
+// update endpoint for the device
+// required:
+// * machineId
+// * endpoint
+function updateEndpoint(deviceId, endpoint, key, dbNumber) {
+  if (!endpoint) {
+    return Promise.reject(new Error('No endpoint provided'));
+  }
+  return checkDeviceId(deviceId, dbNumber)
+  .then(() => redis.hmset(client, `device-${deviceId}`, 'endpoint', endpoint, 'key', key));
+}
+
+function sendNotifications(feature, payload, isNew, dbNumber) {
+  // make payload a string
+  if (payload && Object.keys(payload).length > 0) {
+    payload = JSON.stringify(payload);
+  }
+  return setClient(dbNumber)
+  .then(() => redis.smembers(client, `${feature}-notifications`))
+  .then(devices => redis.smembers(client, 'all-notifications')
```

Comment:
Nit: Could you move this to the following line, so it clearer what the `.then` refers to?
```
.then(devices =>
  redis.smembers(client, 'all-notifications')
  .then(all => devices.concat(all))
)
.then(
```